ingest/anndata_.py

```diff
@@ -114,17 +124,59 @@ def generate_metadata_file(adata, output_name):
             f.write('\t'.join(headers) + '\n')
             f.write('\t'.join(types) + '\n')
         adata.obs.to_csv(output_name, sep="\t", mode="a", header=None, index=True)
+        AnnDataIngestor.compress_file(output_name)
 
     @staticmethod
     def clusterings_to_delocalize(arguments):
         # ToDo - check if names using obsm_keys need sanitization
         cluster_file_names = []
         for name in arguments["obsm_keys"]:
-            cluster_file_names.append(AnnDataIngestor.set_clustering_filename(name))
+            compressed_file = AnnDataIngestor.set_clustering_filename(name) + ".gz"
+            cluster_file_names.append(compressed_file)
         return cluster_file_names
 
     @staticmethod
-    def delocalize_extracted_files(file_path, study_file_id, files_to_delocalize):
+    def compress_file(filename):
+        with open(filename, 'rb') as file_in:
+            compressed_file = filename + '.gz'
+            with gzip.open(compressed_file, 'wb') as file_gz:
+                shutil.copyfileobj(file_in, file_gz)
+        os.remove(filename)
+
+    @staticmethod
+    def generate_processed_matrix(adata):
+        """
+        Generate matrix files with the following file names:
+        h5ad_frag.matrix.processed.mtx
+        h5ad_frag.barcodes.processed.tsv
+        h5ad_frag.features.processed.tsv
+        Gzip files for faster delocalization
```

Comment:
@eweitz by `classic TSV 2nd pass in the same job` do you mean the normal ingest process for flat files?  If so, I would suggest we should not do that.  The sole reason why is that we can parallelize the downstream parse jobs, whereas doing them here would require serially ingesting the extracted files.

Reply:
> by classic TSV 2nd pass in the same job do you mean the normal ingest process for flat files?

Yes, I believe so.  My understanding is that this work splits what was a 1-pass ingest of processed expression AnnData into 2 passes, where the 2nd pass would be normal ingest for flat MTX TSV files.

> I would suggest we should not do that. The sole reason why is that we can parallelize the downstream parse jobs, whereas doing them here would require serially ingesting the extracted files.

Intriguing, I hadn't considered that.  Are there any diagrams for AnnData ingest dataflow?

Here's my understanding of flow for the expected predominant case of ingesting single-AnnData with single process expression matrix, given the current changes in this PR:

Pass 1:
* Fully ingest clustering and metadata
* Extract processed expression from AnnData format, transform it to to 3 MTX TSV files (i.e., MTX bundle)

Pass 2:
* Load processed expression data from MTX bundle to MongoDB

If we expect AnnData files to _often_ have multiple processed gene expression matrices, then a multi-job approach sounds faster via parallelization.  However, if AnnData files _rarely_ have multiple processed matrices, then a multi-job approach seems slower.  Assessing distribution of # processed expression matrices per study seems like it would help clarify each approach's relative performance benefit.

Median times for classic successful ingest:
* Metadata: [~2 minutes](https://mixpanel.com/s/2ZsQq1) (1 per study)
* Clustering: [~1 minute](https://mixpanel.com/s/N7MSh) ([median: 1 / study; P90: 4, P99: 14](https://broadworkbench.atlassian.net/browse/SCP-4559))
* MTX: [~6 minutes](https://mixpanel.com/s/49g8Si) (median, P90, P99 = ?)

Reply:
I don't think we have diagrams - @jlchang do you know of any?  However the logic as it stands right now is as follows:

Pass 1 (extract):
* Extract flat tsv files for clustering, metadata
* Extract mtx/barcodes/features files for processed expression
* Delocalize "fragment" files back to study bucket

Pass 2-n (parse):
* Fully parse all extracted fragment files in parallel (for n files)
* TBD - subsampling, DE ingest

We don't do any actual ingest during the first "extract" phase.  There is also a proposal to run metadata validation in the first extraction job, as this is the most common kind of ingest error, and that way we can prevent spinning up VMs for downstream jobs that we know will fail eventually.  I think this execution will end up being the fastest, especially as the size of the AnnData file grows.